File: ios/Classes/BannerAd/BannerAdLoader.swift
import Foundation
import GoogleMobileAds
import Flutter

/// Protocol for banner ad loader callbacks.
protocol BannerAdLoaderDelegate: AnyObject {
    func adLoader(_ loader: BannerAdLoader, didReceiveBannerAd bannerView: GADBannerView)
    func adLoader(_ loader: BannerAdLoader, didFailWithError error: Error)
}

/// Handles loading banner ads from AdMob.
class BannerAdLoader: NSObject {

    private let adUnitId: String
    let controllerId: String
    private let channel: FlutterMethodChannel
    private let enableDebugLogs: Bool
    private let adSize: GADAdSize

    private var bannerView: GADBannerView?

    weak var delegate: BannerAdLoaderDelegate?
    private var onAdLoadedCallback: ((GADBannerView) -> Void)?

    init(
        adUnitId: String,
        controllerId: String,
        channel: FlutterMethodChannel,
        adSize: GADAdSize,
        enableDebugLogs: Bool = false
    ) {
        self.adUnitId = adUnitId
        self.controllerId = controllerId
        self.channel = channel
        self.adSize = adSize
        self.enableDebugLogs = enableDebugLogs
        super.init()
    }

    /// Sets the callback for when an ad is loaded.
    func setOnAdLoadedCallback(callback: @escaping (GADBannerView) -> Void) {
        onAdLoadedCallback = callback
    }

    /// Loads a banner ad.
    func loadAd() {
        guard let rootVC = rootViewController() else {
            log("ERROR: Could not get root view controller!")
            return
        }

        let bannerView = GADBannerView(adSize: adSize)
        bannerView.adUnitID = adUnitId
        bannerView.rootViewController = rootVC
        bannerView.delegate = self

        self.bannerView = bannerView
        bannerView.load(GADRequest())
    }

    /// Gets the currently loaded banner view.
    func getBannerView() -> GADBannerView? {
        return bannerView
    }

    /// Destroys the loader and releases resources.
    func destroy() {
        bannerView?.delegate = nil
        bannerView = nil
        onAdLoadedCallback = nil
    }

    // MARK: - Private Methods

    private func rootViewController() -> UIViewController? {
        // iOS 13+ uses connectedScenes
        if #available(iOS 13.0, *) {
            return UIApplication.shared.connectedScenes
                .compactMap { ($0 as? UIWindowScene)?.windows.first }
                .first?.rootViewController
        }
        // Fallback for older iOS versions
        return UIApplication.shared.windows.first?.rootViewController
    }

    private func sendEvent(_ method: String, arguments: [String: Any]) {
        DispatchQueue.main.async { [weak self] in
            self?.channel.invokeMethod(method, arguments: arguments)
        }
    }

    private func log(_ message: String) {
        if enableDebugLogs {
            print("[BannerAdLoader][\(controllerId)] \(message)")
        }
    }
}

// MARK: - GADBannerViewDelegate

extension BannerAdLoader: GADBannerViewDelegate {

    func bannerViewDidReceiveAd(_ bannerView: GADBannerView) {
        sendEvent("onAdLoaded", arguments: [
            "controllerId": controllerId
        ])

        onAdLoadedCallback?(bannerView)
        delegate?.adLoader(self, didReceiveBannerAd: bannerView)
    }

    func bannerView(_ bannerView: GADBannerView, didFailToReceiveAdWithError error: Error) {
        log("Banner ad failed to load: \(error.localizedDescription)")

        let nsError = error as NSError
        sendEvent("onAdFailedToLoad", arguments: [
            "controllerId": controllerId,
            "error": error.localizedDescription,
            "errorCode": nsError.code
        ])

        delegate?.adLoader(self, didFailWithError: error)
    }

    func bannerViewDidRecordClick(_ bannerView: GADBannerView) {
        sendEvent("onAdClicked", arguments: [
            "controllerId": controllerId
        ])
    }

    func bannerViewDidRecordImpression(_ bannerView: GADBannerView) {
        sendEvent("onAdImpression", arguments: [
            "controllerId": controllerId
        ])
    }

    func bannerViewWillPresentScreen(_ bannerView: GADBannerView) {
        sendEvent("onAdOpened", arguments: [
            "controllerId": controllerId
        ])
    }

    func bannerViewWillDismissScreen(_ bannerView: GADBannerView) {
    }

    func bannerViewDidDismissScreen(_ bannerView: GADBannerView) {
        sendEvent("onAdClosed", arguments: [
            "controllerId": controllerId
        ])
    }

    func bannerView(_ bannerView: GADBannerView, didReceiveAdValue value: GADAdValue) {
        let valueInDollars = Double(value.value) / 1_000_000.0
        sendEvent("onAdPaid", arguments: [
            "controllerId": controllerId,
            "value": valueInDollars,
            "currencyCode": value.currencyCode
        ])
    }
}

File: android/src/main/kotlin/com/tqc/ads/flutter_admob_native_ads/banner/BannerAdLoader.kt
package com.tqc.ads.flutter_admob_native_ads.banner

import android.content.Context
import android.util.Log
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdSize
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.LoadAdError
import io.flutter.plugin.common.MethodChannel

/**
 * Handles loading banner ads from AdMob.
 */
class BannerAdLoader(
    private val context: Context,
    private val adUnitId: String,
    private val controllerId: String,
    private val channel: MethodChannel,
    private val adSize: AdSize,
    private val enableDebugLogs: Boolean = false
) {
    companion object {
        private const val TAG = "BannerAdLoader"
    }

    private var adView: AdView? = null
    private var onAdLoadedCallback: ((AdView) -> Unit)? = null

    /**
     * Sets the callback for when an ad is loaded.
     */
    fun setOnAdLoadedCallback(callback: (AdView) -> Unit) {
        onAdLoadedCallback = callback
    }

    /**
     * Loads a banner ad.
     */
    fun loadAd() {
        adView = AdView(context)
        adView?.adUnitId = adUnitId
        adView?.setAdSize(adSize)

        adView?.adListener = object : AdListener() {
            override fun onAdLoaded() {
                adView?.let { onAdLoadedCallback?.invoke(it) }
                sendEvent("onAdLoaded", mapOf("controllerId" to controllerId))
            }

            override fun onAdFailedToLoad(error: LoadAdError) {
                log("Banner ad failed to load: ${error.message} (code: ${error.code})")
                sendEvent("onAdFailedToLoad", mapOf(
                    "controllerId" to controllerId,
                    "error" to (error.message ?: "Unknown error"),
                    "errorCode" to error.code
                ))
            }

            override fun onAdClicked() {
                sendEvent("onAdClicked", mapOf("controllerId" to controllerId))
            }

            override fun onAdImpression() {
                sendEvent("onAdImpression", mapOf("controllerId" to controllerId))
            }

            override fun onAdOpened() {
                sendEvent("onAdOpened", mapOf("controllerId" to controllerId))
            }

            override fun onAdClosed() {
                sendEvent("onAdClosed", mapOf("controllerId" to controllerId))
            }
        }

        adView?.loadAd(AdRequest.Builder().build())
    }

    /**
     * Gets the currently loaded banner view.
     */
    fun getAdView(): AdView? = adView

    /**
     * Destroys the loader and releases resources.
     */
    fun destroy() {
        adView?.destroy()
        adView = null
        onAdLoadedCallback = null
    }

    /**
     * Sends an event to Flutter via method channel.
     */
    private fun sendEvent(method: String, arguments: Map<String, Any?>) {
        try {
            channel.invokeMethod(method, arguments)
        } catch (e: Exception) {
            log("Error sending event $method: ${e.message}")
        }
    }

    private fun log(message: String) {
        if (enableDebugLogs) {
            Log.d(TAG, "[$controllerId] $message")
        }
    }
}
